airbyte-integrations/connectors/destination-postgres/src/main/java/io/airbyte/integrations/destination/postgres/typing_deduping/PostgresSqlGenerator.java

```diff
@@ -0,0 +1,294 @@
+import static org.jooq.impl.DSL.array;
+import static org.jooq.impl.DSL.case_;
+import static org.jooq.impl.DSL.cast;
+import static org.jooq.impl.DSL.field;
+import static org.jooq.impl.DSL.function;
+import static org.jooq.impl.DSL.name;
+import static org.jooq.impl.DSL.quotedName;
+import static org.jooq.impl.DSL.rowNumber;
+import static org.jooq.impl.DSL.val;
+
+import com.google.common.collect.ImmutableMap;
+import io.airbyte.cdk.integrations.base.JavaBaseConstants;
+import io.airbyte.cdk.integrations.destination.NamingConventionTransformer;
+import io.airbyte.cdk.integrations.destination.jdbc.TableDefinition;
+import io.airbyte.cdk.integrations.destination.jdbc.typing_deduping.JdbcSqlGenerator;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteProtocolType;
+import io.airbyte.integrations.base.destination.typing_deduping.AirbyteType;
+import io.airbyte.integrations.base.destination.typing_deduping.Array;
+import io.airbyte.integrations.base.destination.typing_deduping.ColumnId;
+import io.airbyte.integrations.base.destination.typing_deduping.Sql;
+import io.airbyte.integrations.base.destination.typing_deduping.StreamConfig;
+import io.airbyte.integrations.base.destination.typing_deduping.Struct;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.jooq.Condition;
+import org.jooq.DataType;
+import org.jooq.Field;
+import org.jooq.SQLDialect;
+import org.jooq.impl.DefaultDataType;
+import org.jooq.impl.SQLDataType;
+
+public class PostgresSqlGenerator extends JdbcSqlGenerator {
+
+  public static final DataType<?> JSONB_TYPE = new DefaultDataType<>(null, Object.class, "jsonb");
+
+  private static final Map<String, String> POSTGRES_TYPE_NAME_TO_JDBC_TYPE = ImmutableMap.of(
+      "numeric", "decimal",
+      "int8", "bigint",
+      "bool", "boolean",
+      "timestamptz", "timestamp with time zone",
+      "timetz", "time with time zone");
+
+  public PostgresSqlGenerator(final NamingConventionTransformer namingTransformer) {
+    super(namingTransformer);
+  }
+
+  @Override
+  public Sql setup() {
+    return Sql.separately(
+        // safe_cast implementation copied from https://dba.stackexchange.com/a/203986
+        // and wrapped in the idempotent function creation logic from https://stackoverflow.com/a/69547602
+
+        // This is pretty hacky. It's to handle the fact that postgres disallows concurrent
+        // `CREATE OR REPLACE FUNCTION airbyte_safe_cast` calls.
+        // It's possible for two syncs to start running at the same time, and we can't coordinate
+        // between destination-postgres instances, so we need to make sure that they can both successfully
+        // run this function creation logic without failing.
+        """
+        do $$
+        begin
+            CREATE FUNCTION airbyte_safe_cast(_in text, INOUT _out ANYELEMENT)
```

Comment:
Does it make sense to add schema qualifier to avoid permission issues in whatever the default schema the function is created ? basically the same raw_schema/raw_schema_override name

Reply:
done in 8969cf954e - took a little refactoring to get the raw table schema in the sqlgenerator